3.0/Dockerfile

```diff
@@ -1,66 +1,90 @@
 
 ENV DEBIAN_FRONTEND noninteractive
 ENV LANG C.UTF-8
 
-RUN apt-get -y update -qq && \
-    apt-get -y install locales && \
-    locale-gen en_US.UTF-8 && \
-    update-locale LANG=en_US.UTF-8 && \
-    apt-get install -y build-essential cmake g++ libboost-dev libboost-system-dev \
-    libboost-filesystem-dev libexpat1-dev zlib1g-dev libxml2-dev\
-    libbz2-dev libpq-dev libgeos-dev libgeos++-dev libproj-dev \
-    postgresql-server-dev-9.5 postgresql-9.5-postgis-2.2 postgresql-contrib-9.5 \
-    apache2 php php-pgsql libapache2-mod-php php-pear php-db \
-    php-intl git curl sudo \
-    python-pip libboost-python-dev \
-    gdal-bin python-gdal unzip \
-    osmosis && \
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/* && \
-    rm -rf /tmp/* /var/tmp/*
+RUN apt-get -y update -qq
+RUN apt-get -y install locales
+RUN locale-gen en_US.UTF-8
+RUN update-locale LANG=en_US.UTF-8
 
-WORKDIR /app
-
-# Configure postgres
-RUN echo "host all  all    0.0.0.0/0  trust" >> /etc/postgresql/9.5/main/pg_hba.conf && \
-    echo "listen_addresses='*'" >> /etc/postgresql/9.5/main/postgresql.conf
+# Install dependencies (automatically removed in multi-stage build)
+RUN apt-get install -y build-essential cmake g++ libboost-dev libboost-system-dev \
+libboost-filesystem-dev libexpat1-dev zlib1g-dev libxml2-dev \
+libpq-dev libgeos-dev libgeos++-dev libproj-dev libbz2-dev \
+postgresql-server-dev-9.5 git
+# Cleanup is not necessary with multi-stage build!  these dependencies will not be copied over!
 
-RUN sed -i -e"s/^max_connections = 100.*$/max_connections = 1000/" /etc/postgresql/9.5/main/postgresql.conf
+WORKDIR /app/src
 
-# Nominatim install
+# Nominatim install and build
 ENV NOMINATIM_VERSION v3.0.1
-RUN git clone --recursive https://github.com/openstreetmap/Nominatim ./src
-RUN cd ./src && git checkout tags/$NOMINATIM_VERSION && git submodule update --recursive --init && \
-    mkdir build && cd build && cmake .. && make
+RUN git clone --recursive https://github.com/openstreetmap/Nominatim .
+RUN git checkout tags/$NOMINATIM_VERSION && git submodule update --recursive --init
+RUN mkdir build && cd build && cmake .. && make
+
+# RUNTIME!
+FROM ubuntu:xenial
+
+# TODO Do we want to drop the /src at this point?
+WORKDIR /app/src
+COPY --from=0 /app/src .
+
+ENV DEBIAN_FRONTEND noninteractive
+ENV LANG C.UTF-8
+
+RUN apt-get -y update -qq
+RUN apt-get -y install locales
+RUN locale-gen en_US.UTF-8
+RUN update-locale LANG=en_US.UTF-8
 
-# Osmium install to run continuous updates
+# Load initial data
+COPY country_osm_grid.sql.gz data/country_osm_grid.sql.gz
+COPY data.osm.pbf data/data.osm.pbf
+ADD tiger.tar.gz data/
+
+# Runtime Dependencies
+RUN apt-get install -y apache2 php php-pgsql libapache2-mod-php php-pear php-db \
+php-intl python-pip libboost-python-dev gdal-bin python-gdal \
+postgresql-9.5-postgis-2.2 postgresql-contrib-9.5 \
+libbz2-dev libz-dev unzip sudo
+
+# TODO Possibly Unnecessary Dependencies?
+RUN apt-get install -y osmosis
```

Comment:
I think this is necessary. Apparently this image can update its data during runtime, and I believe Osmosis is the tool that allows that feature.